Integration/assessment update revised

Kelv1nG opened the pull request:
### 🎫 Ticket

- [integration: assessment update](https://trello.com/c/EQp2I0oh/74-assessments-update-integration)

- use backend branch: https://github.com/Kelv1nG/School-ERP-backend/tree/branch-for-testing-with-assessment-update
### 📦 Packages

- None

### 📝 Updates

main:
- create generic assessment pages for assignment / quiz / exam
- create assessment cubit for managing create/update of assessment page
- updated base crud repository, added tests

minor:
- remove queries / methods from models, moved to repositories for consistency

create:
https://drive.google.com/file/d/1fDB529zp-pGoh2BBcweDIW47vZbdA95E/view?usp=drive_link

update:
https://drive.google.com/file/d/1NET_HTQUDyj8ovFXIFPDQNMX5tlBg_Kj/view?usp=sharing

### 🧪 Testing Steps
- for base crud repo: `flutter test -v .\test\integration\base_repository\base_repository_test.dart`
- only tested assessment service / assessment cubit in app ( no tests yet )

Kelv1nG committed:
rearrange dto directory location

Kelv1nG committed:
add utils for converting bool to int

powersync table has no bool type

Kelv1nG committed:
assessment service

- bulk creation of assessment taker in  create
- implemented update
- added result types

Kelv1nG committed:
create initial result type for assessment service

Kelv1nG committed:
test before changing assessment dtos

Kelv1nG committed:
update assessment pages

Kelv1nG committed:
update repositories

Kelv1nG committed:
model updates

Kelv1nG committed:
change implementation of base_crud_repository

Kelv1nG committed:
Create base_model.dart

Kelv1nG committed:
create crud_base_repository tests

Kelv1nG committed:
remove dtos

Kelv1nG committed:
update assessment service, add assessment result

Kelv1nG committed:
create assessment cubit

cubit + state

Kelv1nG committed:
remove previous assessment service test

Kelv1nG committed:
update base crud repository

Kelv1nG committed:
update repositories

Kelv1nG committed:
create common initialized repositories

Kelv1nG committed:
Update sql_statements.dart

Kelv1nG committed:
Create teacher_repository.dart

Kelv1nG committed:
Update number_input.dart

Kelv1nG committed:
Update assignment_card.dart

Kelv1nG committed:
update models

Kelv1nG committed:
update enums

Kelv1nG committed:
update getTeacherId implementation

Kelv1nG committed:
create generic assessment question setup page

Kelv1nG committed:
create generic assessment setup form

Kelv1nG committed:
create generic assessment takers

Kelv1nG committed:
Update assignment_list_page.dart

ejempic commented on lib/features/assessment/cubit/assessment_cubit.dart:
```diff
@@ -0,0 +1,172 @@
+          isLoading: false,
+        ));
+      } else {
+        // Create a default assessment taker for create mode
+        var defaultTaker = AssessmentTaker(
+          assessmentId: state.assessment.id ?? '',
+          sectionId: '',
+          startTime: DateTime.now().add(const Duration(minutes: 30)),
+          deadLine: DateTime.now().add(const Duration(days: 1)),
+        );
+
+        emit(state.copyWith(
+          assessmentTakers: [defaultTaker],
+          isLoading: false,
+        ));
```
why do we need this?

Kelv1nG replied:
used to render a single assessment taker item when creating an assessment
![image](https://github.com/user-attachments/assets/02d27c49-3b0e-475b-be08-f0a97a93608c)

but yes the '' fields for ids are non ideal, please refer to updated proposal

ejempic commented on lib/features/assessment/cubit/assessment_cubit.dart:
```diff
@@ -0,0 +1,172 @@
+
+  void addAssessmentTaker() {
+    var assessmentTaker = AssessmentTaker(
+      assessmentId: state.assessment.id ?? '',
+      sectionId: '',
```
handle null requirements in object class remove required fields if not actually required. Unsutainable if we duplicate or replicate to another feature

ejempic commented on lib/pages/assessment/assessment_create_update/assessment_question_setup/assesment_question_setup_form.dart:
```diff
@@ -0,0 +1,135 @@
+      );
+    } else {
+      ScaffoldMessenger.of(context).showSnackBar(
+        const SnackBar(content: Text('Please fix the errors in red')),
```
PLease move to constants texts

ejempic commented on lib/pages/assessment/assessment_create_update/assessment_question_setup/assesment_question_setup_form.dart:
```diff
@@ -0,0 +1,135 @@
+                    const SizedBox(height: 5),
+                    const SizedBox(height: 25),
+                    LabeledDropdownFormField<bool>(
+                      label: "AI Generated",
```
Contants assessments.labels.ai_generated

ejempic commented on lib/pages/assessment/assessment_create_update/assessment_question_setup/assesment_question_setup_form.dart:
```diff
@@ -0,0 +1,135 @@
+                    ),
+                    const SizedBox(height: 25),
+                    NumberInputFormField(
+                        label: 'No. of Questions',
```
Contants assessments.labels.no_of_questions

ejempic commented on lib/pages/assessment/assessment_create_update/assessment_question_setup/assesment_question_setup_form.dart:
```diff
@@ -0,0 +1,135 @@
+                    NumberInputFormField(
+                        label: 'No. of Questions',
+                        initialValue: state.assessment.totalQuestions,
+                        helperText: 'Enter the total number of questions',
```
Contants assessments.helpers.total_no_questions

ejempic commented on lib/pages/assessment/assessment_create_update/assessment_question_setup/assesment_question_setup_form.dart:
```diff
@@ -0,0 +1,135 @@
+                        helperText: 'Enter the total number of questions',
+                        validator: (value) {
+                          if (value == null || value.isEmpty) {
+                            return 'Please enter the number of questions';
```
Contants assessments.validations.enter_no_questions

ejempic commented on lib/pages/assessment/assessment_create_update/assessment_question_setup/assesment_question_setup_form.dart:
```diff
@@ -0,0 +1,135 @@
+                          }
+                          final number = int.tryParse(value);
+                          if (number == null) {
+                            return 'Please enter a valid number';
```
Contants general.validations.number

ejempic commented on lib/pages/assessment/assessment_create_update/assessment_question_setup/assesment_question_setup_form.dart:
```diff
@@ -0,0 +1,135 @@
+                            return 'Please enter a valid number';
+                          }
+                          if (number <= 0) {
+                            return 'Number of questions must be greater than 0';
```
Contants general.validations.number_gt_0

ejempic commented on lib/pages/assessment/assessment_create_update/assessment_question_setup/assesment_question_setup_form.dart:
```diff
@@ -0,0 +1,135 @@
+                        }),
+                    const SizedBox(height: 25),
+                    LabeledDropdownFormField<bool>(
+                      label: "Select Randomness",
```
Contants assessments.labels.random

ejempic commented on lib/pages/assessment/assessment_create_update/assessment_setup/assessment_setup_form.dart:
```diff
@@ -0,0 +1,215 @@
+      );
+    } else {
+      ScaffoldMessenger.of(context).showSnackBar(
+        const SnackBar(content: Text('Please fix the errors in red')),
```
same

ejempic commented on lib/pages/assessment/assessment_create_update/assessment_setup/assessment_setup_form.dart:
```diff
@@ -0,0 +1,215 @@
+      initialValue: state.assessment.title,
+      maxLength: 30,
+      decoration: const InputDecoration(
+        hintText: 'What should you call this assignment',
```
Contants assessments.helper.***

ejempic commented on lib/pages/assessment/assessment_create_update/assessment_setup/assessment_setup_form.dart:
```diff
@@ -0,0 +1,215 @@
+      ),
+      validator: (value) {
+        if (value == null || value.isEmpty) {
+          return 'Enter Title';
```
Contants assessments.validation.**

ejempic reviewed (changes requested):
PLeas address pr request changes and use constant text except for thrown error message.

Kelv1nG committed:
update assessment and assessment_taker model

make some id fields nullable but include 2 factory constructors
a) factory ModelName in which the ids are required
b) factory ModelName.initialize where the ids are null

Kelv1nG committed:
update assessment cubit with the updated models

Kelv1nG committed:
update base crud repo

throw exception when no new row is inserted

Kelv1nG committed:
Update assessment_setup_form.dart

Kelv1nG committed:
update comment in assessment cubit for default assessment taker

Kelv1nG committed:
added form constants labels/validation/text hint

Kelv1nG committed:
update pages
